File: src/main/java/com/itsci/mju/maebanjumpen/service/FileSystemStorageService.java
package com.itsci.mju.maebanjumpen.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.util.StringUtils;
import org.apache.commons.io.FilenameUtils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.UUID;

@Service
public class FileSystemStorageService implements StorageService {

    @Value("${file.upload-dir}")
    private String uploadDir;

    @Override
    public String uploadFile(MultipartFile file, String folderName) throws Exception {
        if (file.isEmpty()) {
            throw new Exception("Failed to store empty file.");
        }

        System.out.println("uploadDir: " + uploadDir);
        System.out.println("folderName: " + folderName);

        String originalFilename = StringUtils.cleanPath(file.getOriginalFilename());
        String fileExtension = FilenameUtils.getExtension(originalFilename);
        String newFilename = UUID.randomUUID().toString();
        if (!fileExtension.isEmpty()) {
            newFilename += "." + fileExtension;
        }

        Path uploadPath = Paths.get(uploadDir, folderName).toAbsolutePath().normalize();
        Files.createDirectories(uploadPath);

        Path destinationFile = uploadPath.resolve(newFilename);

        try (InputStream inputStream = file.getInputStream()) {
            Files.copy(inputStream, destinationFile, StandardCopyOption.REPLACE_EXISTING);
            System.out.println("File saved successfully!");
        } catch (IOException e) {
            System.err.println("Error storing file " + originalFilename + " in " + uploadPath + ": " + e.getMessage());
            throw new Exception("Failed to store file " + originalFilename, e);
        }
        return newFilename;
    }

    @Override
    public String uploadFile(MultipartFile file, String folderName, String subfolderName) throws Exception {
        if (file.isEmpty()) {
            throw new Exception("Failed to store empty file.");
        }

        String originalFilename = StringUtils.cleanPath(file.getOriginalFilename());
        String fileExtension = FilenameUtils.getExtension(originalFilename);
        String newFilename = UUID.randomUUID().toString();
        if (!fileExtension.isEmpty()) {
            newFilename += "." + fileExtension;
        }

        // สร้าง Path สำหรับโฟลเดอร์ย่อย
        Path uploadPath = Paths.get(uploadDir, folderName, subfolderName).toAbsolutePath().normalize();
        Files.createDirectories(uploadPath); // สร้างโฟลเดอร์ย่อยถ้ายังไม่มี

        Path destinationFile = uploadPath.resolve(newFilename);

        try (InputStream inputStream = file.getInputStream()) {
            Files.copy(inputStream, destinationFile, StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException e) {
            throw new Exception("Failed to store file " + originalFilename, e);
        }
        return newFilename;
    }

    @Override
    public Path loadFile(String filename, String folderName) {
        return Paths.get(uploadDir, folderName, filename).toAbsolutePath().normalize();
    }

    @Override
    public Path loadFile(String filename, String folderName, String subfolderName) {
        return Paths.get(uploadDir, folderName, subfolderName, filename).toAbsolutePath().normalize();
    }

    @Override
    public byte[] loadFileAsBytes(String filename, String folderName) throws Exception {
        Path filePath = loadFile(filename, folderName);
        if (!Files.exists(filePath) || !Files.isReadable(filePath)) {
            throw new Exception("File not found or not readable: " + filename);
        }
        return Files.readAllBytes(filePath);
    }

    @Override
    public boolean deleteFile(String filename, String folderName) {
        if (filename == null || filename.isEmpty()) {
            return false;
        }
        try {
            Path filePath = Paths.get(uploadDir, folderName, filename).toAbsolutePath().normalize();
            return Files.deleteIfExists(filePath);
        } catch (IOException e) {
            System.err.println("Error deleting file: " + filename + " from folder " + folderName + " - " + e.getMessage());
            return false;
        }
    }

    @Override
    public boolean deleteFile(String filename, String folderName, String subfolderName) {
        if (filename == null || filename.isEmpty()) {
            return false;
        }
        try {
            Path filePath = Paths.get(uploadDir, folderName, subfolderName, filename).toAbsolutePath().normalize();
            return Files.deleteIfExists(filePath);
        } catch (IOException e) {
            System.err.println("Error deleting file: " + filename + " from folder " + subfolderName + " - " + e.getMessage());
            return false;
        }
    }
}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/other/service/impl/FileSystemStorageService.kt
package com.itsci.mju.maebanjumpen.other.service.impl

import com.itsci.mju.maebanjumpen.other.service.StorageService
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.util.*

@Service
class FileSystemStorageService : StorageService {

    @Value("\${app.upload.dir:uploads}")
    private lateinit var uploadDir: String

    override fun uploadFile(file: MultipartFile, folderName: String, subfolderName: String): String {
        try {
            val targetLocation = getTargetLocation(folderName, subfolderName)
            Files.createDirectories(targetLocation)

            val originalFilename = file.originalFilename ?: "file"
            val extension = originalFilename.substringAfterLast('.', "")
            val storedFilename = "${UUID.randomUUID()}.$extension"

            val targetPath = targetLocation.resolve(storedFilename)
            Files.copy(file.inputStream, targetPath, StandardCopyOption.REPLACE_EXISTING)

            return storedFilename
        } catch (e: IOException) {
            throw RuntimeException("Could not store file. Please try again!", e)
        }
    }

    override fun loadFile(filename: String, folderName: String, subfolderName: String): Path {
        val targetLocation = getTargetLocation(folderName, subfolderName)
        return targetLocation.resolve(filename).normalize()
    }

    override fun deleteFile(filename: String, folderName: String, subfolderName: String): Boolean {
        return try {
            val filePath = loadFile(filename, folderName, subfolderName)
            Files.deleteIfExists(filePath)
        } catch (e: IOException) {
            false
        }
    }

    private fun getTargetLocation(folderName: String, subfolderName: String): Path {
        return if (subfolderName == "no-subfolder") {
            Paths.get(uploadDir, folderName)
        } else {
            Paths.get(uploadDir, folderName, subfolderName)
        }
    }
}

